nni/runtime/tuner_command_channel/websocket.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+"""
+Synchronized and object-oriented WebSocket class.
+
+WebSocket guarantees that messages will not be divided at API level.
+"""
+
+from __future__ import annotations
+
+__all__ = ['WebSocket']
+
+import asyncio
+import logging
+from threading import Lock, Thread
+from typing import Any
+
+import websockets
+
+_logger = logging.getLogger(__name__)
+
+# the singleton event loop
+_event_loop: asyncio.AbstractEventLoop = None  # type: ignore
+_event_loop_lock: Lock = Lock()
+_event_loop_refcnt: int = 0  # number of connected websockets
+
+class WebSocket:
+    """
+    A WebSocket connection.
+
+    Call :meth:`connect` before :meth:`send` and :meth:`receive`.
+
+    All methods are thread safe.
+
+    Parameters
+    ----------
+    url
+        The WebSocket URL.
+        For tuner command channel it should be something like ``ws://localhost:8080/tuner``.
+    """
+    def __init__(self, url: str):
+        self._url: str = url
+        self._ws: Any = None  # the library does not provide type hints
+
+    def connect(self) -> None:
+        global _event_loop, _event_loop_refcnt
+        with _event_loop_lock:
```

Comment:
Does this imply that `/tuner` can serve multiple clients simultaneously?

Reply:
No. But a python process can have multiple experiment and each experiment serves a client.